src/update_engine/filesystem_iterator.cc

```diff
@@ -111,11 +111,10 @@ void FilesystemIterator::Increment() {
 void FilesystemIterator::IncrementInternal() {
   CHECK_EQ(dirs_.size(), names_.size() + 1);
   for (;;) {
-    struct dirent dir_entry;
     struct dirent* dir_entry_pointer;
-    int r;
+    errno = 0;
     RETURN_ERROR_IF_FALSE(
-        (r = readdir_r(dirs_.back(), &dir_entry, &dir_entry_pointer)) == 0);
+        (dir_entry_pointer = readdir(dirs_.back())) != NULL || errno != 0);
```

Comment:
I may be misreading this, but I think the check should be `p != NULL || errno == 0`. The error condition we are trying to catch is the `p == NULL && errno != 0`, but negated.

Reply:
Yes, thanks.  Updated.